drgn/helpers/linux/cgroup.py

```diff
@@ -0,0 +1,183 @@
+# SPDX-License-Identifier: GPL-3.0+
+
+"""
+Cgroup
+------
+
+The ``drgn.helpers.linux.cgroup`` module provides helpers for working with
+cgroup interface in :linux:`include/linux/cgroup.h`. Only cgroup v2 is
+supported.
+"""
+
+
+import enum
+
+from drgn import cast, container_of
+from drgn.helpers.linux.kernfs import kernfs_name, kernfs_path
+from drgn.helpers.linux.list import list_for_each_entry
+
+
+__all__ = [
+    'cgroup_name',
+    'cgroup_parent',
+    'cgroup_path',
+    'css_for_each_child',
+    'css_for_each_descendant_pre',
+    'css_next_child',
+    'css_next_descendant_pre',
+    'sock_cgroup_ptr',
+]
+
+
+def sock_cgroup_ptr(skcd):
+    """
+    .. c:function:: struct cgroup *sock_cgroup_ptr(struct sock_cgroup_data *skcd)
+
+    Get cgroup of a socket by ``struct sock_cgroup_data`` member embedded to
+    every ``struct sock``.
+
+
+    :return: ``struct cgroup *`` object.
+    """
+    return cast('struct cgroup *', skcd.val)
+
+
+def cgroup_parent(cgrp):
+    """
+    .. c:function:: struct cgroup *cgroup_parent(struct cgroup *cgrp)
+
+    Get parent of cgroup if it exists.
+
+    :return: ``struct cgroup`` object or None.
+    """
+    parent_css = cgrp.self.parent
+    if parent_css:
+        return container_of(parent_css, 'struct cgroup', 'self')
+    return None
+
+
+def cgroup_name(cgrp):
+    """
+    .. c:function:: char *cgroup_name(struct cgroup *cgrp)
+
+    Get cgroup name.
+
+    :rtype: bytes
+    """
+    return kernfs_name(cgrp.kn)
+
+
+def cgroup_path(cgrp):
+    """
+    .. c:function:: char *cgroup_path(struct cgroup *cgrp)
+
+    Get cgroup path.
+
+    :rtype: bytes
+    """
+    return kernfs_path(cgrp.kn)
+
+
+def _get_css_flags(prog):
+    enum_name = 'CssFlags'
+    try:
+        return prog.cache[enum_name]
+    except KeyError:
+        prog.cache[enum_name] = enum.IntEnum(
+            enum_name, prog['CSS_ONLINE'].type_.enumerators
+        )
+        return prog.cache[enum_name]
+
+
+def css_next_child(pos, parent):
+    """
+    .. c:function:: struct cgroup_subsys_state *css_next_child(struct cgroup_subsys_state *pos, struct cgroup_subsys_state *parent)
+
+    Find the next child of a given css parent starting from the position pos
+    (None to initiate traversal).
+
+    :return: ``struct cgroup_subsys_state *`` object or None.
+    """
+    css_flags = _get_css_flags(parent.prog_)
+
+    if pos is None:
+        next_ = container_of(
+            parent.children.next, 'struct cgroup_subsys_state', 'sibling'
+        )
+    elif not (pos.flags.value_() & css_flags.CSS_RELEASED):
+        next_ = container_of(
+            pos.sibling.next, 'struct cgroup_subsys_state', 'sibling'
+        )
+    else:
+        for next_ in list_for_each_entry(
+            'struct cgroup_subsys_state',
+            parent.children.address_of_(),
+            'sibling'
+        ):
+            if next_.serial_nr.value_() > pos.serial_nr.value_():
```

Comment:
Likewise here, you don't need `.value_()`. Although in this case, it would be a bit more efficient to read `pos.serial_nr` once (under the covers, each access needs to read from `/proc/kcore` or the core dump):

```
pos_serial_nr = pos.serial_nr.value_()
for ...:
    if next_.serial_nr > pos_serial_nr:
        break
```